### PR TITLE
run photo and track scan on app installation

### PR DESCRIPTION
**Pull request by eneiluj**
We can run the scan with an install repair step. I've tested it on a fresh Nextcloud 16 install. I added a few photos and track files, then installed Maps app and it works :smile:.

Downside of this method, i guess an admin can think there is some kind of failure if the installation is too long.

### CODE REVIEW

**Comment by jancborchardt** [comment]
cc @rullzer @MorrisJobke @nickvergessen can you check if this a good way to do it? :) The goal is to check through people’s files for photos and tracks to be displayed on the map directly on first run.

**Comment by rullzer** [comment]
Yeah that won't work. For example if I install the app on my personal instance now it will have to go over 500 gb of pictures.

Also a big chance this will time out on several systems. 

What you could do is insert a background job in the repair step.

This background job then does what you want. And will be run by cron.

Then I would also propose to have some config var that gets set when the initial scan is done. So you can show a message "your initial scan is not complete yet"

**Comment by tacruc** [comment]
@rullzer is timeout a problem for background jobs? So the initial scan be in multiple background jobs or is one big job running multiple hours fine in an background job?

**Comment by rullzer** [comment]
> @rullzer is timeout a problem for background jobs? So the initial scan be in multiple background jobs or is one big job running multiple hours fine in an background job?

It depends. If people run it from cron then it is fine.
If people run it from ajax there is still the timeout problem.

What we usually do in server is a layered system.

1. Repair step inserts backgroundjob 1
2. backgroundjob1 loops over all seen users and for each inserts bacgroundjob2
3. bakgroundjob2 does the actual scanning for that user.

That way we make sure that one huge job doesn't stop the queue from progressing.

**Comment by eneiluj** [comment]
@rullzer Thanks a lot for the advice. I'm on it!

**Comment by eneiluj** [comment]
@rullzer Why not inserting all bakgroundjob2 directly from the install repair step?

**Comment by rullzer** [comment]
> @rullzer Why not inserting all bakgroundjob2 directly from the install repair step?

Consider a system with >20k users

**Comment by eneiluj** [comment]
Yep a'ight :smile:.

**Comment by eneiluj** [comment]
Ok done. The repair step schedules a "job1" which schedules a "job2" for each user. A "job2" scans photos and tracks for a user.

I've tested it with a small amount of users and files. It works with Ajax cron task.

User is warned if the media scan was not done for him/her yet.

**Comment by jancborchardt** [comment]
I disabled the app and enabled it again to test (that works, right?). It shows the notification still:

> Media scan was not done yet. Wait a few minutes/hours and reload this page to see your photos/tracks.

I only have a few photos (15 already shown and I think those are all geotagged ones) and 0 tracks atm. So it should not show the note anymore?

**Comment by eneiluj** [comment]
@jancborchardt It depends the way the cron task is triggered on your NC instance. If it's with Ajax, there might be lots of other jobs to launch before the ones scheduled by Maps. If it's with a system cron job, you have to wait for it to be triggered twice because first there is a job scheduling one job per user and then those jobs are executed.

**Comment by eneiluj** [comment]
You can check if something is scheduled with:

```
sqlite3 /var/www/html/test/nextcloud/data/owncloud.db 'select * from oc_jobs;' | grep -i maps
```

assuming you use SQLite and you have access to the DB file.

File: js/script.js [hunk 1]
```diff
@@ -179,6 +179,14 @@
                 async: true
             }).done(function (response) {
                 optionsValues = response.values;
+
+                // check if install scan was done
+                if (!optionsValues.hasOwnProperty('installScanDone') || optionsValues.installScanDone !== 'yes') {
+                    OC.Notification.showTemporary(
+                        t('maps', 'Media scan was not done yet. Wait a few minutes/hours and reload this page to see your photos/tracks.')
```
Comment (tacruc) [inline js/script.js]: minutes/hours I think just mentions hours is fine here.

**Comment by jancborchardt** [comment]
Ok – @rullzer could you check if the code/mechanism is proper? :)

**Comment by jancborchardt** [comment]
@eneiluj and from the database I get this:
```
321|OCA\Maps\BackgroundJob\LookupMissingGeoJob|[]|0|1566941482|0|0
```
Do the 0s mean it’s done? :D

**Comment by eneiluj** [comment]
It's not the install scan job there. This one is independent and dedicated to contact addresses.

Either the install job was not scheduled when you disabled/enabled the app or it has been executed.

What about

```
sqlite3 /var/www/html/test/nextcloud/data/owncloud.db 'select * from oc_maps_photos;' | wc -l
```

**Comment by eneiluj** [comment]
It's much simpler to test this with a fresh NC install to make sure there is nothing in Maps DB before testing this branch.

**Comment by jancborchardt** [comment]
@eneiluj the other command just outputs "35".

Yeah, as said I might not be the best person to test this as this needs a proper code review. So @rullzer or maybe @ChristophWurst @nickvergessen could briefly check. :)

**Comment by eneiluj** [comment]
@jancborchardt So there are 35 photos found by Maps app. If you started with an empty Nextcloud instance, this means the scan was done after Maps installation. If you tested with existing data in Maps DB, this does not show if install scan works :wink:.

**Comment by eneiluj** [comment]
Thanks @tacruc . I'll still wait a few minutes/hours (:wink:) for a confirmation that I did it properly technically speaking. Then I'll merge and produce the v0.1.0 archive!

**Comment by nickvergessen** [comment]
Looks good in general, but I guess the scanning will fail for bigger users with ajax cron, but well that's their tough luck. Maybe a console command to rescan a user can help in that case.

**Comment by eneiluj** [comment]
Thanks @nickvergessen. There already are two commands, one to scan photos and the other to scan tracks.

File: js/script.js (hunk 1 above)
Comment (jancborchardt) [reply js/script.js]: I’d say let’s keep minutes/hours cause "hours" sounds a bit harsh, no? :D

**Pull request merged**